Feature/210 decentralize did registry

tahpot opened the pull request:
Will close #210.

pranavburnwal commented:
In behest of #213, This PR is outdated. 

cc @ITStar10 @tahpot

tahpot commented:
This is updating did-client

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
-            throw new Error("Unable to save DIDDocument. No private key.")
+    public async saveDocument(document: DIDDocument | undefined): Promise<boolean> {
+        if (!this.didDoc) {
+            return Promise.reject("No original document")
```
It should be possible to save a new document.

If `this.didDoc` doesn't exist, create a new empty document.

ITStar10 replied:
Created an empty document when original document is not existed.

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+    // private did?: string
+    
+    // Verida Doc
+    private didDoc? : DIDDocument
```
DID Client can not store DID documents. Its designed to be called to fetch and save did documents, but not retain any state information.

ITStar10 replied:
Removed.

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+    constructor(config: DIDClientConfig) {
+        this.config = config
+
+        this.identifier = config.identifier
```
This should be calculated from the VDA private key

ITStar10 replied:
Updated.

ITStar10 replied:
There is no verida private key in constructor params now.
Verida private key will be inputed in authenticate() function.
Now in the constructor, did-client interpret the inputed identifiers and deduce DID from it. Input param for `identifier` can be an DID or an address value.

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
     }
 
+    public authenticate(privateKey: string) {
+        throw new Error('Method not implemented.')
```
This method accepts a private key and then unlocks the ability to save documents.

It should be possible to "get" a document, without providing a private key.

ITStar10 replied:
Once privateKey not provided in constructor, there should be an identifier in config. If not we can't create a vda-did-resolver and consequently not able to get document.

ITStar10 replied:
Implemented.

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+                web3: config.web3
+            }
+
+        this.vdaDid = new VdaDID({
```
See below, this should only happen when the user "authenticate()"s.

ITStar10 replied:
Done.

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+
+    getPublicKey(): string;
+
+    saveDocument(document: DIDDocument): Promise<boolean>;
```
Rename to "save()" to match the previous interface to avoid breaking other protocol code

ITStar10 replied:
done

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+
+    saveDocument(document: DIDDocument): Promise<boolean>;
+
+    getDocument(): Promise<DIDDocument | null>
```
Rename to "get()" to match the previous interface to avoid breaking other protocol code

ITStar10 replied:
done

tahpot commented on packages/did-client/src/did-client.ts:
```diff
@@ -1,90 +1,308 @@
+/**
+ * Interface to DIDClient instance
+ */
+export interface DIDClient {
```
This interface needs to match the methods the previous did-client implementation had.

ITStar10 replied:
Will remove interface and export class directly.
This interface was designed to load did document asynchronously in the did-document creation.
Not necessary now.

ITStar10 reviewed (commented):
Should remove unnecessary test files in vda-did-xx libraries.

tahpot commented:
Next steps for @ITStar10 in `vda-did-resolver`:

1. Update the configuration to have `testnet` that uses Verida contract address and chainId. Remove all the ethereum networks.
2. Replace `ethr` with `vda` for all the code, so the library expects `vda` DID's.

`did-document`:

1. @tahpot to produce a broken unit test to show this issue

tahpot commented:
@ITStar10 I have added back the original unit tests in `did-client`.

None of them pass, but we need them to all pass. Can you please review and investigate what is needed to change so they all pass?

ITStar10 commented on packages/did-document/src/did-document.ts:
```diff
@@ -101,10 +103,10 @@ export default class DIDDocument {
         })
         this.doc.assertionMethod = this.doc.assertionMethod!.filter((entry: string | VerificationMethod) => {
-            return entry !== `${this.doc.id}?context=${contextHash}#sign`
+            return entry !== `${this.doc.id}?context=${contextHash}`
```
![image](https://user-images.githubusercontent.com/15656252/192335236-d2004a93-0189-4c12-afb8-ab83b13e40a8.png)
Please take a look for Key purpose of public keys.
As there are purpose of keys while adding by setAttribute() function, we don't need feel to set key type after context.
As you can see, there are 3 key purposes of veriKey, sigAuth, & enc.
Here I matched "#sign" to sigAuth & "#asym" to veriKey.
Do we need to add more key types?
And I am not sure whether adding more types is compatible with DID spec.

ITStar10 replied:
Link is https://github.com/decentralized-identity/ethr-did-resolver/blob/master/doc/did-method-spec.md

tahpot replied:
I have resolved this issue with this commit: https://github.com/verida/verida-js/pull/211/commits/b069bd590456a95c9134cd04f23994583fd545ed

tahpot replied:
Actually, I haven't resolved this.

How can `did-document` specify the key purpose so that it is saved? I think that's the problem. This code needs to set the key purpose so that `vda-did-resolver` can save it to the blockchain.

ITStar10 replied:
Please take a look line#93 ~ line#100 of did-client/src/helpers.ts.
You can know that how it deduce the keyPurpose from created DIDDocument.
For each item in the document.verificationMethod, if same id is contained in document.authentication the keyPurpose is 'sigAuth'. If same id is contained in document.keyAgreemtn, keyPurpose is 'enc'.

![image](https://user-images.githubusercontent.com/15656252/193190373-f959c8f5-21d2-44bd-8dc5-4e5a8cd5d9cb.png)

tahpot replied:
How does this map back to the DID document when the smart contract is read?

Can we make this more explicit by using query parameters in the publicKey that specify “typ”? That logic will not always hold true in the future

ITStar10 replied:
Please check line#266 ~ line#270 in the vda-did-resolver/src/resolver.ts. Same logic. I created above code from the vda-did-resolver. They're matched.

![image](https://user-images.githubusercontent.com/15656252/193191839-485fca17-5805-4a70-ae33-d8004cca14de.png)

ITStar10 replied:
Of course, we can add type in the attribute value. To do so, need to check out vda-did & vda-did-resolver again.

tahpot commented:
I have merged #231 which solves the following issues:

1. The code now expects `vda` instead of `ethr` in the did registry client library
2. Original `did-client` unit tests have been added back and now pass
3. Default did registry configuration uses testnet and previous ethereum networks have been removed

The following issues remain:

1. There are lots of commented out tests in `did-client.basic.test.ts`. Do you want to remove these or uncomment them?
2. `chainId` being inserted into `did` (this is not part of the VDA spec and will break everything)
3. Inability to handle both `asym` and `sym` public keys for an application context

ITStar10 commented:
- did-client.basic.test.ts file is already removed.
- chainId is for DID-Registry-contracts deployed on the blockchains other than Polygon.
If we're aiming at Polygon only, we can remove chainId in the configuration. chainId is used to create RPCProvider.
Instead of chainId in the configuration, we can use only chainName and we should have a list of chain IDs of supported chain name in our code.
- As mentioned before, please check `addContextSignKey()` and `addContextAsymKey()` functions in the did-document.ts.
- Will check storage-link.

ITStar10 commented:
> There are lots of commented out tests in did-client.basic.test.ts. Do you want to remove these or uncomment them?
This file is already removed.

tahpot commented:
@ITStar10 I have updated the did-client tests to show how they don't work.

See [this commit](https://github.com/verida/verida-js/pull/211/commits/eebf84ec550d8e19a2c32580e9ad144bafff2c94)

I have updated the test to do a comparison on the original DID document that was to be saved and then on the DID document retrieved from the blockchain.

These are the issues:

1. `verificationMethod`: `type=asym` and `type=sign` values are not being saved. `context` not being saved.
2. `keyAgreement`: `type=asym` is not being saved. `context` not being saved.
3. `assertionMethod`: `type=sign` is not being saved. `context` not being saved.
